File: VodArchiver/VodArchiver/VideoJobs/YoutubeVideoJob.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Xml;
using VodArchiver.VideoInfo;

namespace VodArchiver.VideoJobs {
	public class YoutubeVideoJob : IVideoJob {
		public YoutubeVideoJob( string id, StatusUpdate.IStatusUpdate statusUpdater = null ) {
			JobStatus = VideoJobStatus.NotStarted;
			StatusUpdater = statusUpdater == null ? new StatusUpdate.NullStatusUpdate() : statusUpdater;
			VideoInfo = new GenericVideoInfo() { Service = StreamService.Youtube, VideoId = id };
			Status = "...";
		}

		public YoutubeVideoJob( XmlNode node ) : base( node ) { }

		public override bool IsWaitingForUserInput => false;

		public override XmlNode Serialize( XmlDocument document, XmlNode node ) {
			node.AppendAttribute( document, "_type", "YoutubeVideoJob" );
			return base.Serialize( document, node );
		}

		public override async Task<ResultType> Run( CancellationToken cancellationToken ) {
			if ( cancellationToken.IsCancellationRequested ) { return ResultType.Cancelled; }

			JobStatus = VideoJobStatus.Running;
			if ( ( VideoInfo as YoutubeVideoInfo ) == null ) {
				Status = "Retrieving video info...";
				var result = await Youtube.RetrieveVideo( VideoInfo.VideoId, VideoInfo.Username );

				switch ( result.result ) {
					case Youtube.RetrieveVideoResult.Success:
						VideoInfo = result.info;
						break;
					case Youtube.RetrieveVideoResult.ParseFailure:
						// this seems to happen randomly from time to time, just retry later
						return ResultType.TemporarilyUnavailable;
					default:
						return ResultType.Failure;
				}
			}

			string filenameWithoutExtension = "youtube_" + VideoInfo.Username + "_" + VideoInfo.VideoTimestamp.ToString( "yyyy-MM-dd" ) + "_" + VideoInfo.VideoId;
			string filename = filenameWithoutExtension + ".mkv";
			string tempFolder = Path.Combine( Util.TempFolderPath, filenameWithoutExtension );
			string tempFilepath = Path.Combine( tempFolder, filename );

			{
				if ( !await Util.FileExists( tempFilepath ) ) {
					if ( cancellationToken.IsCancellationRequested ) { return ResultType.Cancelled; }

					Directory.CreateDirectory( tempFolder );
					Status = "Running youtube-dl...";
					await StallWrite( tempFilepath, 0, cancellationToken ); // don't know expected filesize, so hope we have a sensible value in minimum free space
					if ( cancellationToken.IsCancellationRequested ) { return ResultType.Cancelled; }
					List<string> args = new List<string>() {
						"-f", "bestvideo+bestaudio/best",
						"-o", tempFilepath,
						"--merge-output-format", "mkv",
						"--no-color",
						"--abort-on-error",
						"--abort-on-unavailable-fragment",
					};
					string limit = Util.YoutubeSpeedLimit;
					if ( limit != "" ) {
						args.Add( "--rate-limit" );
						args.Add( limit );
					}
					args.Add( "https://www.youtube.com/watch?v=" + VideoInfo.VideoId );
					var data = await ExternalProgramExecution.RunProgram(
						@"youtube-dl", args.ToArray(),
						stdoutCallbacks: new System.Diagnostics.DataReceivedEventHandler[1] {
							( sender, received ) => {
								if ( !String.IsNullOrEmpty( received.Data ) ) {
									Status = received.Data;
								}
							}
						}
					);
				}

				string finalFilename = GenerateOutputFilename();
				string finalFilepath = Path.Combine( Util.TargetFolderPath, finalFilename );
				if ( File.Exists( finalFilepath ) ) {
					throw new Exception( "File exists: " + finalFilepath );
				}

				Status = "Waiting for free disk IO slot to move...";
				try {
					await Util.ExpensiveDiskIOSemaphore.WaitAsync( cancellationToken );
				} catch ( OperationCanceledException ) {
					return ResultType.Cancelled;
				}
				try {
					// sanity check
					Status = "Sanity check on downloaded video...";
					TimeSpan actualVideoLength = ( await FFMpegUtil.Probe( tempFilepath ) ).Duration;
					TimeSpan expectedVideoLength = VideoInfo.VideoLength;
					if ( actualVideoLength.Subtract( expectedVideoLength ).Duration() > TimeSpan.FromSeconds( 5 ) ) {
						// if difference is bigger than 5 seconds something is off, report
						Status = "Large time difference between expected (" + expectedVideoLength.ToString() + ") and actual (" + actualVideoLength.ToString() + "), stopping.";
						return ResultType.Failure;
					}

					Status = "Moving...";
					await Task.Run( () => Util.MoveFileOverwrite( tempFilepath, finalFilepath ) );
					await Task.Run( () => Directory.Delete( tempFolder ) );
				} finally {
					Util.ExpensiveDiskIOSemaphore.Release();
				}
			}

			Status = "Done!";
			JobStatus = VideoJobStatus.Finished;
			return ResultType.Success;
		}

		public override string GenerateOutputFilename() {
			return "youtube_" + VideoInfo.Username + "_" + VideoInfo.VideoTimestamp.ToString("yyyy-MM-dd") + "_" + VideoInfo.VideoId + "_" + Util.MakeIntercapsFilename(VideoInfo.VideoTitle).Crop(80) + ".mkv";
		}
	}
}

File: native/vodarchiver/videojobs/youtube-video-job.cpp
#include "youtube-video-job.h"

#include <format>
#include <memory>
#include <string>
#include <string_view>
#include <utility>

#include "util/file.h"
#include "util/text.h"

#include "vodarchiver/exec.h"
#include "vodarchiver/ffmpeg_util.h"
#include "vodarchiver/filename_util.h"
#include "vodarchiver/videoinfo/youtube-video-info.h"
#include "vodarchiver/youtube_util.h"

namespace VodArchiver {
bool YoutubeVideoJob::IsWaitingForUserInput() const {
    return false;
}

static std::string PathCombine(std::string_view lhs, std::string_view rhs) {
    std::string result(lhs);
    HyoutaUtils::IO::AppendPathElement(result, rhs);
    return result;
}

static std::string GetTargetFilename(IVideoInfo& vi) {
    return std::format("youtube_{}_{}_{}_{}.mkv",
                       vi.GetUsername(),
                       DateToString(vi.GetVideoTimestamp()),
                       vi.GetVideoId(),
                       Crop(MakeIntercapsFilename(vi.GetVideoTitle()), 80));
}

static ResultType RunYoutubeVideoJob(YoutubeVideoJob& job,
                                     JobConfig& jobConfig,
                                     TaskCancellation& cancellationToken) {
    std::unique_ptr<IVideoInfo> vi;
    bool wantCookies = false;
    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.JobStatus = VideoJobStatus::Running;
        vi = job.VideoInfo->Clone();
        wantCookies = (job.Notes.find("cookies") != std::string::npos);
    }

    std::string tempFolderPath;
    std::string targetFolderPath;
    {
        std::lock_guard lock(jobConfig.Mutex);
        tempFolderPath = jobConfig.TempFolderPath;
        targetFolderPath = jobConfig.TargetFolderPath;
    }

    if (cancellationToken.IsCancellationRequested()) {
        return ResultType::Cancelled;
    }

    if (dynamic_cast<YoutubeVideoInfo*>(vi.get()) == nullptr) {
        {
            std::lock_guard lock(jobConfig.Mutex);
            job.TextStatus = "Retrieving video info...";
        }
        auto result = Youtube::RetrieveVideo(vi->GetVideoId(), vi->GetUsername(), wantCookies);

        switch (result.result) {
            case Youtube::RetrieveVideoResult::Success:
                vi = result.info->Clone();
                {
                    std::lock_guard lock(jobConfig.Mutex);
                    job.VideoInfo = std::move(result.info);
                }
                break;
            case Youtube::RetrieveVideoResult::ParseFailure:
                // this seems to happen randomly from time to time, just retry later
                return ResultType::TemporarilyUnavailable;
            default: return ResultType::Failure;
        }
    }

    std::string filenameWithoutExtension = std::format("youtube_{}_{}_{}",
                                                       vi->GetUsername(),
                                                       DateToString(vi->GetVideoTimestamp()),
                                                       vi->GetVideoId());
    std::string filename = filenameWithoutExtension + ".mkv";
    std::string tempFolder = PathCombine(tempFolderPath, filenameWithoutExtension);
    std::string tempFilepath = PathCombine(tempFolder, filename);

    {
        if (!HyoutaUtils::IO::FileExists(std::string_view(tempFilepath))) {
            if (cancellationToken.IsCancellationRequested()) {
                return ResultType::Cancelled;
            }

            if (!HyoutaUtils::IO::CreateDirectory(std::string_view(tempFolder))) {
                return ResultType::Failure;
            }
            {
                std::lock_guard lock(jobConfig.Mutex);
                job.TextStatus = "Running youtube-dl...";
            }
            // don't know expected filesize, so hope we have a sensible value in minimum free space
            StallWrite(jobConfig,
                       tempFilepath,
                       0,
                       cancellationToken,
                       ShouldStallWriteRegularFile,
                       [&](std::string status) {
                           std::lock_guard lock(jobConfig.Mutex);
                           job.TextStatus = std::move(status);
                       });
            if (cancellationToken.IsCancellationRequested()) {
                return ResultType::Cancelled;
            }
            std::vector<std::string> args = {{
                "-f",
                "bestvideo[height<=?1080]+bestaudio/best",
                "-o",
                tempFilepath,
                "--merge-output-format",
                "mkv",
                "--no-color",
                "--abort-on-error",
                "--abort-on-unavailable-fragment",
                "--no-sponsorblock",
            }};
            if (wantCookies) {
                args.push_back("--cookies");
                args.push_back("d:\\cookies.txt");
            }
            args.push_back("https://www.youtube.com/watch?v=" + vi->GetVideoId());
            std::string output;
            if (RunProgram(
                    "yt-dlp.exe",
                    args,
                    [&](std::string_view sv) {
                        output.append(sv);

                        // only keep the last line
                        std::string_view trimmed = HyoutaUtils::TextUtils::Trim(output);
                        auto pos = trimmed.find_last_of("\r\n");
                        if (pos != std::string_view::npos) {
                            output.assign(trimmed.substr(pos + 1));
                        }

                        {
                            std::lock_guard lock(*jobConfig.JobsLock);
                            job.TextStatus = output;
                        }
                    },
                    [](std::string_view sv) {})
                != 0) {
                return ResultType::Failure;
            }
        }

        std::string finalFilename = GetTargetFilename(*vi);
        std::string finalFilepath = PathCombine(targetFolderPath, finalFilename);
        if (HyoutaUtils::IO::Exists(std::string_view(finalFilepath))) {
            std::lock_guard lock(*jobConfig.JobsLock);
            job.TextStatus = ("File exists: " + finalFilepath);
            return ResultType::Failure;
        }

        {
            std::lock_guard lock(*jobConfig.JobsLock);
            job.TextStatus = "Waiting for free disk IO slot to move...";
        }
        {
            auto diskLock = jobConfig.ExpensiveDiskIO.WaitForFreeSlot(cancellationToken);
            if (cancellationToken.IsCancellationRequested()) {
                return ResultType::Cancelled;
            }

            // sanity check
            {
                std::lock_guard lock(*jobConfig.JobsLock);
                job.TextStatus = "Sanity check on downloaded video...";
            }
            auto probe = FFMpegProbe(tempFilepath);
            if (!probe) {
                std::lock_guard lock(*jobConfig.JobsLock);
                job.TextStatus = "Probe failed!";
                return ResultType::Failure;
            }
            TimeSpan actualVideoLength = probe->Duration;
            TimeSpan expectedVideoLength = vi->GetVideoLength();
            if (std::abs((actualVideoLength - expectedVideoLength).GetTotalSeconds()) > 5.0) {
                // if difference is bigger than 5 seconds something is off, report
                std::lock_guard lock(*jobConfig.JobsLock);
                job.TextStatus = std::format(
                    "Large time difference between expected ({}s) and actual ({}s), stopping.",
                    expectedVideoLength.GetTotalSeconds(),
                    actualVideoLength.GetTotalSeconds());
                return ResultType::Failure;
            }

            {
                std::lock_guard lock(*jobConfig.JobsLock);
                job.TextStatus = "Moving...";
            }
            if (!HyoutaUtils::IO::Move(
                    std::string_view(tempFilepath), std::string_view(finalFilepath), true)) {
                return ResultType::Failure;
            }
            if (!HyoutaUtils::IO::DeleteDirectory(std::string_view(tempFolder))) {
                return ResultType::Failure;
            }
        }
    }

    {
        std::lock_guard lock(*jobConfig.JobsLock);
        job.TextStatus = "Done!";
        job.JobStatus = VideoJobStatus::Finished;
    }
    return ResultType::Success;
}

ResultType YoutubeVideoJob::Run(JobConfig& jobConfig, TaskCancellation& cancellationToken) {
    return RunYoutubeVideoJob(*this, jobConfig, cancellationToken);
}

std::string YoutubeVideoJob::GenerateOutputFilename() {
    auto vi = GetVideoInfo();
    return GetTargetFilename(*vi);
}
} // namespace VodArchiver
